training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala

```diff
@@ -237,21 +238,22 @@ object AdditiveModelTrainer {
     // transform example as-is to initialize model which we can base our vector compression upon
     val transformed = (frac: Double) => LinearRankerUtils.makePointwiseFloat(input(frac), transformer)
     var model = modelInitialization(sc, transformed, params)
+    var modelBC = sc.broadcast(model.generateFeatureIndexer())
     var bestModel = model
 
-    val Array(trainVectorRDD, validationVectorRDD) = transformFeatureVectorAndSplit(sc, input, params, transformer, model)
+    // broadcast a separate copy of model because the modelBC will be destroyed after each iteration
+    val Array(trainVectorRDD, validationVectorRDD) = transformFeatureVectorAndSplit(sc, input, params, transformer, sc.broadcast(model))
 
     var learningRate = params.learningRate.initialLearningRate
     var trainingLosses = List(Double.MaxValue)
     var validationLosses = List(Double.MaxValue)
     var bestValidationLoss = Double.MaxValue
     var bestIteration = 0
     val sgdParams = SgdParams(params, sc.accumulator(0), sc.accumulator(0))
-    var modelBC = sc.broadcast(model.generateFeatureIndexer())
     var i = 0
     while (i < iterations &&
       (params.loss.earlyStopping == 0 || (i - bestIteration) < params.loss.earlyStopping) &&
-      (trainingLosses.length < 2 || (trainingLosses(1) - trainingLosses(0) > params.loss.convergenceTolerance))
+      (trainingLosses.length < 2 || params.loss.convergenceTolerance <= 0 || (trainingLosses(1) - trainingLosses(0) > params.loss.convergenceTolerance))
```

Comment:
When you don't want to stop iteration even when training loss might be increasing. This can happen if training and evaluation loss is quite correlated. More iterations might still help because learning rate decay will kick in and could possibly further decrease loss down the road. Those cases are much better addressed using early stopping, which stops iterations early when loss has not improved for certain number of iterations. `convergenceTolerance` is useful if you want to experiment things quickly to get a "good enough" model.